### ISSUE TITLE
ccd should be true by default

### ISSUE THREAD

**Issue by JeremyTubongbanua** [atsign-foundation/at_server#1004]
## Summary

ccd (cascade delete) should be set to true by default.

If I share data with somebody and decide to unshare it, the cached key does not cascade delete (because it was set to false by default). They can still see my data even though I've deleted the key. The cached key lives on their secondary server, so in a way, they do own my data. And I do not have any authority over the cached key so I cannot "unshare" my sensitive data.

## Solutions/Alternatives

1. set ccd to true by default
2. for every refresh, make cached keys delete themselves if they see if the original key was deleted.

## Example

A more in-depth explanation of the problem at hand.
```
1. @alice decides to share her ssn with @bob
```@alice@update:ttr:100:ccd:false:@bob:ssn@alice myssn-123-456 => data:2```
*important to note, ttr > 0 so it will cache if looked up

2. @bob sees this data and since ttr > 0, it will cache
``` @bob@scan => data:[cached:@bob:ssn@alice]```

3. @alice does not want @bob to see her ssn anymore, so @alice deletes the key
```@alice@delete:@bob:ssn@alice => data:3```

4. @bob still has his cached key and can still see @alice's ssn
```@bob@scan => data:[cached:@bob:ssn@alice]```
```@bob@lookup:@bob:ssn@alice => data:myssn-123-456```

5. @alice is angry because her ssn lives on @bob's secondary even though she deleted it. this all happened because the developer forgot to set ccd to true.
```

**Comment by gkc** [comment]
Assigned 2 SP for this to get discussed on Arch call

**Comment by sitaram-kalluri** [comment]
Discussion from the Arch call: Set the CCD (Cascade Delete) to true by default; which currently is set to false by default.

**Comment by gkc** [comment]
@JeremyTubongbanua will you be able to pick this up in the next sprint?

**Comment by JeremyTubongbanua** [comment]
> @JeremyTubongbanua will you be able to pick this up in the next sprint?

Not sure

UMass is still my focus until December.. also have some other commitments.

Will CCD be set to true at the Protocol level or in the SDKs?

@gkc

**Comment by gkc** [comment]
@JeremyTubongbanua ACK - no rush on this. Not sure I understand your question properly, but for sure we need to change it in the SDKs.

**Comment by JeremyTubongbanua** [comment]
> @JeremyTubongbanua ACK - no rush on this. Not sure I understand your question properly, but for sure we need to change it in the SDKs.

I feel the ccd should be set to true by default through the protocol as well

**Comment by gkc** [comment]
> > @JeremyTubongbanua ACK - no rush on this. Not sure I understand your question properly, but for sure we need to change it in the SDKs.
> 
> I feel the ccd should be set to true by default through the protocol as well

Yep I saw the words but I don't understand what setting it to true "through the protocol" means.

**Comment by JeremyTubongbanua** [comment]
@gkc 

Creating a shared key creates a metadata associated with the key. The default metadata has `ccd:false` 

```
@smoothalligator@ update:ttl:10000:@fascinatingsnow:test@smoothalligator some value
  => data:12534
```

```
@smoothalligator@ llookup:meta:@fascinatingsnow:test@smoothalligator
  => data:{
  "createdBy": null,
  "updatedBy": null,
  "createdAt": "2022-11-15 15:37:58.088Z",
  "updatedAt": "2022-11-15 15:37:58.088Z",
  "availableAt": "2022-11-15 15:37:58.088Z",
  "expiresAt": "2022-11-15 15:38:08.088Z",
  "refreshAt": null,
  "status": "active",
  "version": 0,
  "ttl": 10000,
  "ttb": 0,
  "ttr": null,
  "ccd": false,
  "isBinary": false,
  "isEncrypted": false,
  "dataSignature": null,
  "sharedKeyEnc": null,
  "pubKeyCS": null,
  "encoding": null
  }

```

**Comment by gkc** [comment]
Indeed. So setting the default to true in the SDK has the desired effect. But what do you mean by, "through the protocol"?

**Comment by JeremyTubongbanua** [comment]
@gkc Sorry, hopefully this clears it up:

If the `ccd:false` is not provided when doing `update:<>`, then set `ccd:true` by default.

Example:

1. Create a new SharedKey entirely.
```
@ update:@bob:test@alice some value
``` 
Important detail to note: we did not enter `ccd:true`. Instead of doing `update:ccd:true:@bob:test@alice some value`, we just ran `update:@bob:test@alice some value`

2. Look at the metadata
Even though we did not explicitly say `ccd:true`, the `ccd` is set to `true` by default by the protocol:
```
@ llookup:meta:@bob:test@alice
=> {
  "createdBy": null,
  "updatedBy": null,
  "createdAt": "2022-11-15 16:39:57.658Z",
  "updatedAt": "2022-11-15 16:39:57.658Z",
  "availableAt": "2022-11-15 16:39:57.658Z",
  "expiresAt": null,
  "refreshAt": null,
  "status": "active",
  "version": 0,
  "ttl": 0,
  "ttb": 0,
  "ttr": null,
  "ccd": true,
  "isBinary": false,
  "isEncrypted": false,
  "dataSignature": null,
  "sharedKeyEnc": null,
  "pubKeyCS": null,
  "encoding": null
}
```

I feel it would be beneficial to set `ccd` to `true` by default at the protocol level (instead of through the SDKs) because it would be more reliable, especially when some keys would be holding sensitive data.

What do you think? @gkc

**Comment by gkc** [comment]
So ... via the client-side verb builders? And/or the server-side verb handlers?

**Comment by JeremyTubongbanua** [comment]
@gkc I was thinking no client-side verb builders at all would be involved with making `ccd:true`.
The server-side creating the metadata JSON object would handle setting `ccd:true` by default. Although I would not know how that would work. 

I can still do it for both of our SDKs if that is what you would advise.
